### ISSUE TITLE
pool: task from previous run not retrieved from database

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#5952]
A niche situation where outputs from task which was previously run are not injected into the pool on restart.

This is a cut-down version of the complex example in https://github.com/cylc/cylc-flow/pull/5947:

```ini
#!Jinja2

{% set stop_cycle = stop_cycle | default(4) %}

[scheduler]
    allow implicit tasks = True

[scheduling]
    initial cycle point = 1
    stop after cycle point = {{ stop_cycle }}
    cycling mode = integer
    [[graph]]
        # main sequence
        P1 = a
        # transition
        R/{{ stop_cycle - 2 }} = a => b
        # spin-down sequence
        R2/P1/{{ stop_cycle }} = """
            a => b
            b[-P1] => b
        """
```

To replicate:

```bash
# run to the end of cycle 4
cylc vip -s stop_cycle=4

# then, extend on to cycle 6
cylc play -s stop_cycle=6
```

The workflow will stall due to `4/b:succeeded`:

```
WARNING - Partially satisfied prerequisites:
      * 5/b is waiting on ['4/b:succeeded']
      * 6/b is waiting on ['5/b:succeeded']
```

However, this task had succeeded in the initial run:

```
INFO - [4/b running job:01 flows:1] => succeeded
```

It should have been re-inserted into the pool on restart which would have resulted in its succeeded output satisfying its downstream task `5/b`.

#### Overview in diagrams:

`cylc vip -s stop_cycle=4`

![Screenshot from 2024-01-31 11-25-17](https://github.com/cylc/cylc-flow/assets/16705946/1d152a64-342d-4430-ba7a-d8b3c56a584a)

`cylc play -s stop_cycle=6`

![Screenshot from 2024-01-31 11-25-28](https://github.com/cylc/cylc-flow/assets/16705946/69be7a2b-573f-4257-9d28-65b37a88f49d)

**Comment by oliver-sanders** [comment]
Suggest investigating with https://github.com/cylc/cylc-flow/pull/5658 as there have been a lot of changes in this part of the code.

**Comment by hjoliver** [comment]
I think I agree that we *should* make this work, but it's not a slam dunk.

From an event-driven spawn-on-demand perspective, the restart has a different graph that does not reflect what actually happened before shutdown.

First, all the green tasks finish complete, and we shut down with `5/a` waiting in the task pool:
![gr1](https://github.com/cylc/cylc-flow/assets/2362137/d90cc76f-4551-4579-b8e2-29bfab01ee02)

Then, we demand a restart (so, load `5/a` ready to run) with a later stop-cycle, but change the graph to this:
![gr2](https://github.com/cylc/cylc-flow/assets/2362137/0d02aabd-8a59-4a78-bb2a-38dd3fc07c9f)

The dependency `4/b => 5/b` did not exist when `4/b` completed and its children (in this case, none) were spawned.

**Comment by hjoliver** [comment]
>    `* 5/b is waiting on ['4/b:succeeded']`

Unfortunately the stall message makes this look like a rather trivial bug, but from an event-driven perspective it really means this:

>  ` * the dependence of 5/b on 4/b:succeeded is not satisfied`
>  `   (either 4/b did not succeed OR this dependence did not exist when it did)`

We just have to educate users that dependencies get satisfied automatically *when the corresponding outputs are generated*.

Of course we could in principle handle this automatically by going to the db outputs table, but (without thinking it through yet) that might entail a nasty performance cost just to handle rare runtime graph structure changes. 

I presume we would not want to look up parent outputs every time a task gets spawned. 

I suppose we could do it on stall, just for partially satisfied tasks. In this case, that would automatically unstall the workflow.

**Comment by oliver-sanders** [comment]
> The reason for "pool: task from previous run not retrieved from database" is that, from an event-driven spawn-on-demand perspective, under normal circumstances (i.e., if you don't change the structure of the graph in the active window mid-run) we should never have to do that.

Sorry, but I don't accept this argument. If this were true, then we wouldn't perform DB checks when we insert tasks:

https://github.com/cylc/cylc-flow/blob/8ab57ad0943abeea5b297f7d3b19d3cd8a055175/cylc/flow/task_pool.py#L1503-L1506

https://github.com/cylc/cylc-flow/blob/8ab57ad0943abeea5b297f7d3b19d3cd8a055175/cylc/flow/task_pool.py#L1580-L1582

Cylc tells the user that a task which has previously run and succeeded, has not:

```
WARNING - Partially satisfied prerequisites:
      * 5/b is waiting on ['4/b:succeeded']
```

And that is unarguably a bug!

It doesn't matter whether it's "right" by the internal logic of SoD, it matters that is is "wrong" by the task/dependency model of Cylc - https://github.com/cylc/cylc-flow/issues/2143

**Comment by hjoliver** [comment]
@oliver-sanders - au contraire! 

> I understand what you are saying from the perspective of the SoD implementation, but I don't accept this argument from the perspective of the Cylc model

The Cylc 8 scheduling model is event-driven, and the current behaviour is perfectly understandable in that context, as I've shown above. You changed the graph mid-run to add a new dependency on an event that happened in the past. 

There's no law of nature that says we have any right to expect event-driven systems to work like that to work out of the box. 
If we can do it (by looking up past events), then I have already agreed that we should, but that's an addition to the model, not a bug fix. 

From your other comments I think perhaps you haven't analysed the graph correctly and as a result you think there is a bug in how we handle spawning at the stop point?? - but that isn't the case - see below.

> Note, if this were true, then we wouldn't perform DB checks when we insert tasks:

That's a different check. It looks up the submit and flow history of the task being spawned (by a current event).  For the example we need to look up the outputs of all the parent tasks of the one being spawned, just in case any of those outputs were generated in the past at a time **when this task (the one being spawned) did not depend on them**!

> Moreover, Cylc tells the user that a task which has previously run and succeeded, has not, both in the log:

See my comments above on that. The wording is entirely up to us, of course, but even now it does not actually say that - it just says, correctly, that the task is still waiting on that output. And why is it still waiting? Because that new dependency did not exist back in the past when the output was generated.

Message wording pedantry aside, I do agree that it will *look that way* to users, and (again) I haven't said we should not do this, just explained why the current behavior is an expected consequence (not a bug) of the current conceptual (not implementation) scheduling model. 

> Note also that this is not actually a graph change.  

Sorry, that's inarguably false - just look at my graphs above!!  Your original graphs at the top are wrong - they don't show the critical inter-cycle dependencies.

> The "stop after cycle point" is merely a flag that tells the scheduler when to shut down, the recurrence is not terminated until the "final cycle point" so this dependency does conceptually exist in the first run,

Ah, no - the graph structure here is literally determined (via Jinja2 code) by the current value of the stop point. Here is the complete graph for the initial run, with a stop cycle of 4 and final cycle 10 (note the glaring absence of the problem dependency):

![image](https://github.com/cylc/cylc-flow/assets/2362137/63ea5467-510f-436c-abeb-7a28908b4ac6)

> the problem occurs because the workflow doesn't spawn tasks beyond this point so the satisfied prereq is not recorded in the DB.  I.E. this behaviour is an SoD implementation choice and not really "correct" even by SoD logic.

I can only think you still have your original incorrect graphs in mind. The scheduler DOES spawn beyond the stop point exactly as it should according to the original graph: `5/a:waiting` ends up in the task pool table ready for restart. There is nothing else that should be there, given that graph, for a stop point of 4.

This is not a bug in spawning tasks beyond the stop point, it's the result of changing the graph mid-run to add a new dependency on a past event.

**Comment by oliver-sanders** [comment]
If we agree to go ahead with checking prereqs for all tasks on spawn, then this issue will be superseded by #6143.

**Comment by oliver-sanders** [comment]
We are now in agreement that we should handle this situation, marking this issue as superseeded by #6143 that will check for prereqs on task spawn ensuring the DB and task_pool remain in sync.

**Comment by hjoliver** [comment]
All good, although for the record I never disagreed that we should handle this, I just disagreed that it was a bug. In other words I stand by my initial comment above on this issue:

> This isn't a bug, although we might still want to handle it automatically for convenience.

Supporting automatic satisfaction of *newly added dependence on past events* is a choice. It probably is what users would expect (all the time?) but it may have performance consequences (which also affects users).
